### ISSUE TITLE
Inside out context

### ISSUE THREAD

**Issue by ramelen** [typst/typst#4665]
### Description

A block that is evaluated before the rest of the document with context on by default, which gets unwrapped at the end so the rest of the document can use the results. This could be a dedicated function like `context` is or it could be something like allowing you to see into contextual variables from an import file.

### Use Case

a `context` block allows the inside code to pull information from the outside code, but not the other way around. This is great for things like numbered headings where a small piece of content looks at the surrounding document. 

However, it's much harder to do something like set the x-height of the body text to be 6 pt. You currently have to use context to find the x-height, then wrap the entire body text in a context block to access the x-height variable. 

A dedicated function would find the x-height in a default environment and then unwrap the context at the end to be used in the rest of the document. This would be code that can't pull information from the outside, but the outside code can access it.

**Comment by laurmaedje** [comment]
I don't really understand what exactly you want to achieve with this, that cannot currently be achieved. Could you elaborate on that?

**Comment by PgBiel** [comment]
I think they mean something like

```typ
#let x-height-ctx = context {
   measure([x]).height
}

#let x-height = x-height-ctx.eval-with(some-initial-or-handpicked-context())
```

That is, executing a context with predetermined values for styles and other things, which wouldn't require placing the context in the document to evaluate what's inside it. Something like freeze state but more flexible.

(At least, that's the only way I can see this being feasible - without fixed values, you'd be implicitly depending on the e.g. current font size, which would require placing in the document, no way around it.)

**Comment by laurmaedje** [comment]
If that's indeed the idea, then it's a dupe of https://github.com/typst/typst/issues/3930

**Comment by PgBiel** [comment]
> If that's indeed the idea, then it's a dupe of #3930

It's worth specifying that, with this proposal, `some-initial-or-handpicked-context()` would work without any prior context. That is, you wouldn't have to wrap your whole document in `context { ... }` to execute the contextual code - you could use this at any point in evaluation, as you'd be able to handpick the styles and other information provided by the context (and/or opt into using just the "default" styles provided by Typst itself). Basically, the idea is to be able to use a function like `measure` with e.g. a predefined font size without needing some outer document context, since that particular operation doesn't actually need any information from the context other than said predefined info (plus default styles).

In other words, it's something like #3930 plus a context constructor.

**Comment by ramelen** [comment]
> In other words, it's something like #3930 plus a context constructor.

I think this sums it up nicely. How hard would something like this be to implement?

**Comment by laurmaedje** [comment]
Would that really be that useful? The x-height also depends on the font in use after all.

And more importantly: Is there really a problem with wrapping almost everything in context? I know that there are some bugs with it currently, but those can be fixed. I think it's not actually that problematic.

**Comment by ramelen** [comment]
> And more importantly: Is there really a problem with wrapping almost everything in context? I know that there are some bugs with it currently, but those can be fixed. I think it's not actually that problematic.

Hmm. I'll have to experiment with it some more I guess. I'll come back if I find anything interesting.

**Comment by laurmaedje** [comment]
I'll close this for now. If you do find anything interesting, please post it and we can reconsider things!
